### PR TITLE
Simple examples

### PR DESCRIPTION
**Pull request by Divesh-Otwani**
This adds a top-level simple-examples folder with a few (right now, just 2) files.

The goal of these examples is be like a tutorial, showing a user how to use linear-base (and a bit of linear types in general). The file IO example from the paper is here (and the mutable array one might be added shortly).

Also, these are meant to be suuuper simple so that someone who's just learning haskell can feel comfy with linear types and not intimidated by them.

### CODE REVIEW

File: simple-examples/Examples.hs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import System.Exit
+
+main :: IO ()
+main = exitSuccess
```
Comment (aspiwack) [inline simple-examples/Examples.hs]: This is a curious idiom. Maybe document.

File: simple-examples/FileIO.hs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+interface, we have code that type checks but will error or cause errors at
+runtime.
+
+-}
+
+
+module FileIO where
+
+import Prelude
+import qualified System.IO as Sys
```
Comment (aspiwack) [inline simple-examples/FileIO.hs]: I would prefer `System` to `Sys`.

File: simple-examples/FileIO.hs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import Control.Monad ()
+import Data.Text
+
+-- Linear Base Imports
+import qualified Control.Monad.Linear.Builder as Control
+import Data.Unrestricted.Linear
+import System.IO.Resource
```
Comment (aspiwack) [inline simple-examples/FileIO.hs]: This should probably be imported qualified. _e.g._ as `Linear` for the sake of this file.

File: simple-examples/FileIO.hs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+printFirstLine :: Sys.FilePath -> Sys.IO ()
+printFirstLine fpath = do
+  fileHandle <- Sys.openFile fpath Sys.ReadMode
+  firstLine <- Sys.hGetLine fileHandle
+  Sys.putStrLn firstLine
+  Sys.hClose fileHandle
+
+
+-- This compiles but can cause issues!
```
Comment (aspiwack) [inline simple-examples/FileIO.hs]: What issues?

File: simple-examples/FileIO.hs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+printFirstLineNoClose :: Sys.FilePath -> Sys.IO ()
+printFirstLineNoClose fpath = do
+  fileHandle <- Sys.openFile fpath Sys.ReadMode
+  firstLine <- Sys.hGetLine fileHandle
+  Sys.putStrLn firstLine
+
+
+-- This compiles, but will error!
```
Comment (aspiwack) [inline simple-examples/FileIO.hs]:
```suggestion
-- This compiles, but will throw an error!
```

File: simple-examples/FileIO.hs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    Note that the current approach is limited.
+    We have to make the continuation use the unit type.
+
+    Enabling a more generic approach with a type index
+    for the multiplicity, as descibed in the paper is a work in progress.
+    This will hopefully result in using
+
+    `(>>==) RIO 'Many a #-> (a -> RIO p b) #-> RIO p b`
+
+    as the non-linear bind operation.
```
Comment (aspiwack) [inline simple-examples/FileIO.hs]: Add a link to the relevant issue.

File: simple-examples/FileIO.hs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+-- i.e., (a #-> RIO b)
+(>>#=) :: RIO a #-> (a #-> RIO b) #-> RIO b
+(>>#=) = (>>=)
+  where
+    Control.Builder{..} = Control.monadBuilder
+
+-- | Non-linear bind
+-- Notice the continuation has a non-linear arrow,
+-- i.e., (() -> RIO b)
+(>>==) :: RIO () #-> (() -> RIO b) #-> RIO b
```
Comment (aspiwack) [inline simple-examples/FileIO.hs]: A more general type would be

```suggestion
(>>==) :: RIO (Unrestricted a) #-> (a -> RIO b) #-> RIO b
```

You are not really making use of this function, though.

File: simple-examples/Pure.hs [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+------------------------------------------------------------
+
+{-
+   A linear function simply "consumes/uses" it's argument exactly once.
+
+   Creating a formal definition is quite tricky and requires a good amount
+   of formalizing Haskell. A simple way to think about it is if (f :: a
+   #-> b), then as (f x) evaluates, x and/or the linear components of x,
+   are syntactically present exactly once, and if in a data structure, is
+   a linear argument to a constructor.
```
Comment (aspiwack) [inline simple-examples/Pure.hs]: I'm not sure what this means.

File: simple-examples/Pure.hs [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+   Consider the next function as an example.
+-}
+
+nonLinearSubsume :: (a,a) -> (a,a)
+nonLinearSubsume (x,_) = (x,x)
+
+{-
+   Notice that the function above could not have a linear arrow. If it
+   did, it would not compile. Why is this?  Well, the first argument is
```
Comment (aspiwack) [inline simple-examples/Pure.hs]:
```suggestion
   did, it would not compile. Why is this?  Well, the first argument would be
```

File: simple-examples/Pure.hs [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+{-
+   Of course, the fact that a function is linear makes no difference in
+   non-linear functions. So, non-linear functions can call linear
+   functions willy-nilly and they will work as expected. To state the
+   obvious, linear functions are regular functions but not all functions
+   are linear functions.
+-}
+
+
+(#.) :: (b #-> c) -> (a #-> b) -> (a #-> c)
```
Comment (aspiwack) [inline simple-examples/Pure.hs]: You could even have all the arrows be linear  here.

File: simple-examples/Pure.hs [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+{-
+   For clarity, we include an example of using such linear constructors.
+   Here, linearHoldExtract must use the inner component linearly.
+   Therefore, it's impossible to implement the following function:
+
+   linearHoldPair :: LinearHolder a #-> (a,a)
+   linearHoldPair (LinearHolder x) = (x,x)
+
+   In fact, we have the following equivalency:
```
Comment (aspiwack) [inline simple-examples/Pure.hs]:
```suggestion
   In fact, we have the following equivalence:
```

(it's vastly more common a term in mathematics and computer science)

File: simple-examples/Pure.hs [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+promote :: (a -> b) -> (ForcedUnlinear a #-> b)
+promote f (ForcedUnlinear x) = f x
+
+{-
+   Another way of saying this is the following equivalence proven by the
+   two functions above:
+
+   (ForcedUnlinear a  #-> b) ≅ (a -> b)
+
+   In the linear haskell POPL '18 paper, this datatype is called
```
Comment (aspiwack) [inline simple-examples/Pure.hs]:
```suggestion
   In the Linear Haskell POPL '18 paper, this datatype is called
```

(it's also called so in linear-base)

File: simple-examples/Pure.hs [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+   UnRestricted.  (It's renamed here for clarity within this context.)
```
Comment (aspiwack) [inline simple-examples/Pure.hs]:
```suggestion
   Unrestricted.
```

File: simple-examples/FileIO.hs (hunk 7 above)
Comment (Divesh-Otwani) [reply simple-examples/FileIO.hs]: Hmm ... yes, that's way better.

Comment (Divesh-Otwani) [reply simple-examples/FileIO.hs]: Though, for the simplicity of this example, I'm not going to use that. (I'd have to fmap a `a #-> Unrestricted a` over `RIO ()` after using `hClose` and I want to keep things incredibly simple.

File: simple-examples/Pure.hs (hunk 10 above)
Comment (Divesh-Otwani) [reply simple-examples/Pure.hs]: Oh yes, I purposely made it simpler.

File: linear-base.cabal [hunk 14]
```diff
@@ -71,8 +71,7 @@ test-suite examples
   other-modules:
     Foreign.List
     Foreign.Heap
-  build-depends:
-    base,
+  build-depends: base,
```
Comment (mboes) [inline linear-base.cabal]: Spurious unrelated change.

File: simple-examples/FileIO.hs [hunk 15]
```diff
@@ -0,0 +1,174 @@
+{-# OPTIONS_GHC -fno-warn-name-shadowing #-}
```
Comment (mboes) [inline simple-examples/FileIO.hs]: Run this file through Ormolu to make the formatting consistent, and remove the redundant whitespace. Ormolu will choke on the linear arrows, because it doesn't support them, but you can easily enough preprocess/postprocess the source files and output as a workaround. cc @mrkkrp

File: simple-examples/Examples.hs (hunk 1 above)
Comment (mboes) [reply simple-examples/Examples.hs]: A cleaner choice would be to use Cabal internal libraries. See https://www.haskell.org/cabal/users-guide/developing-packages.html#sublibs.

File: simple-examples/FileIO.hs (hunk 15 above)
Comment (Divesh-Otwani) [reply simple-examples/FileIO.hs]: Done.

**Comment by Divesh-Otwani** [comment]
@mboes I made those edits & made Pure.hs a bit cleaner. For now I'm going to keep these in /examples in a `Simple` sub-folder.
